File: lib/android/app/src/main/java/com/reactnativenavigation/options/LayoutNode.kt
package com.reactnativenavigation.options

import org.json.JSONObject
import java.util.*

class LayoutNode(val id: String?, val type: Type?, val data: JSONObject, val children: List<LayoutNode?>) {
    enum class Type {
        Component, ExternalComponent, Stack, BottomTabs, SideMenuRoot, SideMenuCenter, SideMenuLeft, SideMenuRight, TopTabs
    }

    internal constructor(id: String?, type: Type?) : this(id, type, JSONObject(), ArrayList<LayoutNode?>()) {}

    val options: JSONObject
        get() = data.optJSONObject("options")

}

File: lib/src/adapters/__mocks__/LayoutNode.ts

export default class LayoutNode {
    readonly type: 'Component' | 'ExternalComponent' | 'Stack' | 'BottomTabs' | 'TopTabs' | 'SideMenuRoot' | 'SideMenuLeft' | 'SideMenuRight' | 'SideMenuCenter' | 'SplitView';
    readonly nodeId: string;
    readonly data: object;
    readonly children: LayoutNode[];

    constructor(layout: any) {
        this.type = layout.type;
        this.nodeId = layout.id;
        this.data = layout.data;
        this.children = layout.children.map((childLayout: any) => new LayoutNode(childLayout));
    }

    toString() {
        let childrenString = '';
        if (this.children.length)
            childrenString = `: [${this.children.map((layoutNode: LayoutNode) => layoutNode.toString())}]`;

        return `${this.nodeId}${childrenString}`;
    }
}